### DISCUSSION TITLE
ONNX export of custom trainset has weird shape

### DISCUSSION THREAD

**Discussion by Angele** [ultralytics/yolov5#4054] [Q&A] [answered]
Can somebody please explain to me what the outputs are ?
![image](https://user-images.githubusercontent.com/1502237/126079472-12386f22-b189-4b5c-91cb-a1b9af14a6ca.png)
Custom trained and exported as .onnx. on Windows.
py ./yolov5/export.py --weights ./weights/best.pt --img 832 --simplify

I got that the 39 are classes + 5, even tough i have no idea why +5.
I cannot find docs that explain the 104,104, 52,52 or 26,26.
But i'm especially confused by the number 42588, since it makes repositories like
https://github.com/BobLd/YOLOv4MLNet/tree/yolo-v5-incl for me impossible to incorporate.

I find that first "output" contains info, the other 3 contain same info splitted (gues makes sense with concat as last layer) that has to go throgh a sigmoid function first.
I was looking around for 1 day, with no luck. Trying to get information out of export.py and detect.py.

Can please somebody with a little more insight state what information (x,y,w,h,c probabilities) they contain and most importantly at what position? How is the number 42588 built ?

**Comment by glenn-jocher** [comment]
@Angele the Netron view at the outputs is very clear, you might want to take a look there instead of at the inputs.

**Reply by Angele** [reply]
Thanks, but i don't seem to get more information out of this tool, than the info you see on the top image there.
Did i miss a feature of this tool ?
I knew when i took the screenshot i should scroll down to avoid this type of answer. My bad.

**Comment by Angele** [comment]
In case you meant looking at the .pt model:
![image](https://user-images.githubusercontent.com/1502237/126081052-e816b6cd-9f2a-4bc4-8ce3-f3b8b4505478.png)
I don't think that sheds a lot of light onto the 42588.

**Reply by glenn-jocher** [reply]
@Angele view ONNX model outputs with Netron

**Reply by glenn-jocher** [reply]
@Angele and yes, if you are interested in ONNX outputs looking at a *.pt model won't do anything for you

**Reply by glenn-jocher** [reply]
@Angele the ONNX output makes it very obvious how the output shapes are created:
<img width="1253" alt="Screenshot 2021-07-19 at 02 16 27" src="https://user-images.githubusercontent.com/26833433/126086722-ffaf827d-3f57-4237-9abb-c2d493035ca3.png">

**Reply by glenn-jocher** [reply]
@Angele the steps to reproduce are:
```
python export.py --simplify
```
Then view with Netron. If you are having problems with Netron then raise them directly on the Netron repository.

**Reply by Angele** [reply]
> @Angele the steps to reproduce are:
> 
> ```
> python export.py --simplify
> ```
> 
> Then view with Netron. If you are having problems with Netron then raise them directly on the Netron repository.

We missunderstood each other. My plan was not to analize the model and improve it, but just get the position of the data from the output. At what position can i find what.

Since it was stated that
Input { 1, 416, 416, 3 } and outputs { 1, 52, 52, 3, 85 }, { 1, 26, 26, 3, 85 }, { 1, 13, 13, 3, 85 }
and the YoloV5 model (YOLOV5l) is shaped:
Input { 1, 3, 640, 640 } and outputs { 1, 3, 80, 80, 19 }, { 1, 3, 40, 40, 19 }, { 1, 3, 20, 20, 19 }

and my numbers differ so much, i was confused.
What Neutron wasn't telling me is it's [pred_count][stats + class_pred] and not for example [class_pred + stats]

**Comment by Angele** [comment, marked as answer]
The shape is predictions times (stats + class prediction) .
Yolo5s makes 25200 predictions while Yolo5l makes 42588 by design.
Hence the [42588, (x, y, full_width, full_heigh, confidence, .... class predictions)]

So to get top left corner of your first rect make: ( output[0, 0] - output [0, 2] / 2 )
To see what prediction for class 0 is do output[0, 5]
To see if an object is there get output[0, 4]

The other outputs, like 104,104, 52,52 or 26,26 for yolo5l can be ignored, if you don't want to build your own yolo.
